Duplicate `extern crate` in doc tests (2021-11-21 Nightly Regression)

We have a Rust-2015 doc test which has an `extern crate $CRATE_DOC_TEST_IS_IN` which started failing to build on nightly 2021-11-21 (but passes on stable/2021-11-20).

It now fails with

```
---- src/chain/transaction.rs - chain::transaction::TransactionData (line 22) stdout ----
error[E0259]: the name `lightning` is defined multiple times
 --> src/chain/transaction.rs:24:1
  |
3 | extern crate lightning;
  | ----------------------- previous import of the extern crate `lightning` here
4 | 
5 | extern crate r#lightning;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^ `lightning` reimported here
  |
  = note: `lightning` must be defined only once in the type namespace of this module

error: aborting due to previous error
```

--- a/compiler/rustc_parse/src/parser/diagnostics.rs
+++ b/compiler/rustc_parse/src/parser/diagnostics.rs
@@ -1121,7 +1121,7 @@ impl<'a> Parser<'a> {
         Ok(P(T::recovered(Some(QSelf { ty, path_span, position: 0 }), path)))
     }
 
-    pub(super) fn maybe_consume_incorrect_semicolon(&mut self, items: &[P<Item>]) -> bool {
+    pub fn maybe_consume_incorrect_semicolon(&mut self, items: &[P<Item>]) -> bool {
         if self.eat(&token::Semi) {
             let mut err = self.struct_span_err(self.prev_token.span, "expected item, found `;`");
             err.span_suggestion_short(

--- a/src/librustdoc/doctest.rs
+++ b/src/librustdoc/doctest.rs
@@ -1,4 +1,4 @@
-use rustc_ast::{self as ast, token};
+use rustc_ast as ast;
 use rustc_data_structures::fx::{FxHashMap, FxHashSet};
 use rustc_data_structures::sync::Lrc;
 use rustc_errors::{ColorConfig, ErrorReported, FatalError};
@@ -537,6 +537,7 @@ crate fn make_test(
             use rustc_errors::emitter::{Emitter, EmitterWriter};
             use rustc_errors::Handler;
             use rustc_parse::maybe_new_parser_from_source_str;
+            use rustc_parse::parser::ForceCollect;
             use rustc_session::parse::ParseSess;
             use rustc_span::source_map::FilePathMapping;
 
@@ -572,9 +573,9 @@ crate fn make_test(
                 }
             };
 
-            match parser.parse_mod(&token::Eof) {
-                Ok((_attrs, items, _span)) => {
-                    for item in items {
+            loop {
+                match parser.parse_item(ForceCollect::No) {
+                    Ok(Some(item)) => {
                         if !found_main {
                             if let ast::ItemKind::Fn(..) = item.kind {
                                 if item.ident.name == sym::main {
@@ -606,10 +607,16 @@ crate fn make_test(
                             break;
                         }
                     }
+                    Ok(None) => break,
+                    Err(mut e) => {
+                        e.cancel();
+                        break;
+                    }
                 }
-                Err(mut e) => {
-                    e.cancel();
-                }
+
+                // The supplied slice is only used for diagnostics,
+                // which are swallowed here anyway.
+                parser.maybe_consume_incorrect_semicolon(&[]);
             }
 
             // Reset errors so that they won't be reported as compiler bugs when dropping the

--- a/src/test/rustdoc-ui/auxiliary/empty-fn.rs
+++ b/src/test/rustdoc-ui/auxiliary/empty-fn.rs
@@ -1,0 +1,3 @@
+// no-prefer-dynamic
+#![crate_type = "lib"]
+pub fn empty() {}

--- a/src/test/rustdoc-ui/issue-91134.rs
+++ b/src/test/rustdoc-ui/issue-91134.rs
@@ -1,0 +1,14 @@
+// compile-flags: --test --crate-name=empty_fn --extern=empty_fn --test-args=--test-threads=1
+// aux-build:empty-fn.rs
+// check-pass
+// normalize-stdout-test: "src/test/rustdoc-ui" -> "$$DIR"
+// normalize-stdout-test "finished in \d+\.\d+s" -> "finished in $$TIME"
+// edition:2021
+
+/// <https://github.com/rust-lang/rust/issues/91134>
+///
+/// ```
+/// extern crate empty_fn;
+/// empty_fn::empty();
+/// ```
+pub struct Something;

--- a/src/test/rustdoc-ui/issue-91134.stdout
+++ b/src/test/rustdoc-ui/issue-91134.stdout
@@ -1,0 +1,6 @@
+
+running 1 test
+test $DIR/issue-91134.rs - something (line 10) ... ok
+
+test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in $TIME
+

--- a/src/test/rustdoc-ui/issue-91134.stdout
+++ b/src/test/rustdoc-ui/issue-91134.stdout
@@ -1,6 +1,6 @@
 
 running 1 test
-test $DIR/issue-91134.rs - something (line 10) ... ok
+test $DIR/issue-91134.rs - Something (line 10) ... ok
 
 test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in $TIME
 

